### PR TITLE
Allow simple arithmetic in number literals

### PR DESCRIPTION
**Pull request by kerams**
Implements https://github.com/fsharp/fslang-suggestions/issues/539.
[RFC](https://github.com/fsharp/fslang-design/blob/main/RFCs/FS-1133-arithmetics-in-literals.md)

`+` (unary and binary), `-` (unary and binary), `*`, `/`, `%`, `&&&`, `|||`, `<<<`, `>>>` are now allowed for integers in literals as well as attributes. `not`, `&&`, `||` are allowed for bools.

nwWZVw56UN

### CODE REVIEW

**Comment by En3Tho** [comment]
Nice. Now the hardest part - an RFC :D.

**Comment by kerams** [comment]
@auduchinok, no, that definitely won't work without parser changes. I'd prefer not to tackle that in this PR anyway.

File: src/Compiler/TypedTree/TypedTreeOps.fs
Comment (En3Tho) [inline src/Compiler/TypedTree/TypedTreeOps.fs:9741]: Maybe this can be SpecificBinopExpr too?

Comment (kerams) [reply src/Compiler/TypedTree/TypedTreeOps.fs:9741]: Don't think so, there's some extra logic there
image

**Comment by T-Gro** [comment]
LGTM , provided tests are added.

**Comment by kerams** [comment]
I wanted to include support for `not`, `||` and `&&`, but even `true || false` is represented as a match expression, so I am not sure if it's worth going to the trouble of trying to reduce that.

**Comment by kerams** [comment]
In addition to what is included in the RFC, logical AND and OR are also supported. A side-effect of the fact that those operations are represented as match expressions is the ability to also use simple `if` expressions.

image

**Comment by vzarytovskii** [comment]
A shower thought - should this go under language flag? It's technically changes compiler flow.

**Comment by kerams** [comment]
@edgarfgp, I've tried

```fsharp
let [<Literal>] switch2 =
    match true with
    | true -> x
    | _ -> false
```

and it doesn't work.

`| Expr.Match (decision = TDSwitch (input = input; cases = [ TCase (DecisionTreeTest.Const (Const.Bool test), TDSuccess ([], targetNum)) ]); targets = [| TTarget (_, t0, _); TTarget (_, t1, _) |]) ->` is a condition tailor-made for how logical operations are represented in the TAST, and I don't think that we necessarily want to create some complex expression evaluator here (that's almost TP territory).

**Review by T-Gro — DISMISSED** [review]

**Comment by vzarytovskii** [comment]
@T-Gro @0101 @KevinRansom this is a **non-breaking** new language feature, however it changes the codegen. Shall this be put under preview version, or we're fine enabling it for all versions?

**Comment by kerams** [comment]
You can do `let [<Literal>] x = 4 ||| 3` today, so you could say this is just an extension of the existing mechanism. Up to you of course.

**Comment by vzarytovskii** [comment]
> `let [<Literal>] x = 4 ||| 3` is valid today, so you could say this is just an extension of the existing mechanism. Up to you of course.

Yeah, it's just changing the codegen, not sure if someone can already rely on it in some tooling/when using quotations, etc.

**Comment by kerams** [comment]
https://github.com/dotnet/fsharp/blob/23e22eadf193cdd7e38ea4fa68c0f76d1b14ca9b/src/Compiler/TypedTree/TypedTreeOps.fs#L9667

One thing I'm not sure about is this comment.

**Comment by Happypig375** [comment]
We should allow at least all operations from C# for parity.
image

**Comment by kerams** [comment]
Decimals are a no-go at the moment, because you can't even do `let [<Literal>] x = 1m`.

**Comment by Happypig375** [comment]
Should also allow conversion operators between numeric types.

**Comment by kerams** [comment]
Plenty of other things could theoretically be added - comparison and equality operators, even match expressions. However, I don't want to stray away from what has been proposed and approved too much, and make things unnecessarily complicated.

**Review by T-Gro — DISMISSED** [review]

**Comment by T-Gro** [comment]
I am for merging this in as-is without preview version restriction and letting others built on top of it with new ideas as needed in separate PRs.

**Comment by auduchinok** [comment]
> I am for merging this in as-is without preview version restriction

Why? It contradicts the whole point of having the language versions. It's not a bug fix, and new things can now be compiled that a previous language version didn't allow.

**Comment by vzarytovskii** [comment]
> > I am for merging this in as-is without preview version restriction
> 
> Why? It contradicts the whole point of having the language versions. It's not a bug fix, and new things can now be compiled that a previous language version didn't allow.

Yep, was thinking the same.

We should hide it behind preview version.

**Comment by T-Gro** [comment]
> > I am for merging this in as-is without preview version restriction
> 
> Why? It contradicts the whole point of having the language versions. It's not a bug fix, and new things can now be compiled that a previous language version didn't allow.

I see it as an extension of a feature that was incomplete.

That being said, it might be bad for developers having different SDK versions between each other (or between them and CI) while using an older F# language version. => you are right.

**Comment by kerams** [comment]
Bump. Does it need Don's blessing?

**Comment by vzarytovskii** [comment]
> Bump. Does it need Don's blessing?
> 
> I have a working draft for expressions in enum definitions, but this PR is a requirement.

I think we're fine with merging it

**Review by vzarytovskii — APPROVED** [review]

**Review by T-Gro — APPROVED** [review]

**Pull request merged**
